src/interfaces/records/messages/records-upload.ts

```diff
@@ -0,0 +1,516 @@
+import type { BaseMessage } from '../../../core/types.js';
+import type { MessageStore } from '../../../store/message-store.js';
+import type { GeneralJws, SignatureInput } from '../../../jose/jws/general/types.js';
+import type { RecordsAttestationPayload, RecordsAuthorizationPayload, RecordsUploadCompleteDescriptor, RecordsUploadCompleteMessage, RecordsUploadPartDescriptor, RecordsUploadPartMessage, RecordsUploadStartDescriptor, RecordsUploadStartMessage } from '../types.js';
+
+import { Encoder } from '../../../utils/encoder.js';
+import { GeneralJwsSigner } from '../../../jose/jws/general/signer.js';
+import { Jws } from '../../../utils/jws.js';
+import { Message } from '../../../core/message.js';
+import { ProtocolAuthorization } from '../../../core/protocol-authorization.js';
+import { removeUndefinedProperties } from '../../../utils/object.js';
+
+import { authorize, validateAuthorizationIntegrity } from '../../../core/auth.js';
+import { Cid, computeCid } from '../../../utils/cid.js';
+import { DwnInterfaceName, DwnMethodName, DwnStateName } from '../../../core/message.js';
+
+type RecordsUploadMessageVariant = RecordsUploadCompleteMessage | RecordsUploadPartMessage | RecordsUploadStartMessage;
+
+type RecordsUploadDescriptorVariant = RecordsUploadCompleteDescriptor | RecordsUploadPartDescriptor | RecordsUploadStartDescriptor;
+
+type RecordsUploadOptions = {
+  protocol?: string;
+  schema?: string;
+  recipient?: string;
+  authorizationSignatureInput: SignatureInput;
+  attestationSignatureInputs?: SignatureInput[];
+};
+
+export type RecordsUploadStartOptions = RecordsUploadOptions & {
+  recordId?: string;
+  dataFormat: string;
+};
+
+export type RecordsUploadPartOptions = RecordsUploadOptions & {
+  recordId: string;
+  index: number;
+  data?: Uint8Array;
+  dataCid?: string;
+  dataSize?: number;
+};
+
+export type RecordsUploadCompleteOptions = RecordsUploadOptions & {
+  recordId: string;
+  count: number;
+  dataCid: string;
+  dataSize: number;
+};
+
+export class RecordsUpload extends Message {
```

Comment:
havent written any tests for this yet.  wanted to get the PR up so folks could give feedback on the API :)